<center><u><h1>Assignment-5: Artificial Neural Networks_ANN</center></u></h1>



<br>

Artificial neural networks replicate the human biological neurons (nerve cells).
![image52.png](https://i.pinimg.com/originals/78/13/30/781330a4105ba005e0aac6b0c0e2a43f.png)

The types of Artificial Neural Networks?<br>
1. Feedforward Neural Network   [FNN]<br>
2. Recurrent Neural Network     [RNN]<br>
3. Convolutional Neural Network [CNN]<br>

**<u>Artificial Neural Networks [ANN]</u>**:<br>

Artificial Neural Networks contain artificial neurons which are called units. These units are arranged in a series of layers that together constitute the whole Artificial Neural Networks in a system. A layer can have only a dozen units or millions of units as this depends on the complexity of the system. Commonly, Artificial Neural Network has an input layer, output layer as well as hidden layers. The input layer receives data from the outside world which the neural network needs to analyze or learn about. Then this data passes through one or multiple hidden layers that transform the input into data that is valuable for the output layer. Finally, the output layer provides an output in the form of a response of the Artificial Neural Networks to input data provided. 

## Types of activation Functions?
Activation function decides, whether a neuron should be activated or not by calculating weighted sum and further adding bias with it. The purpose of the activation function is to introduce non-linearity into the output of a neuron.It is used to determine the output of neural network like yes or no. It maps the resulting values in between 0 to 1 or -1 to 1 etc.

The Activation Functions can be based on 2 types-
1. Linear Activation Function
2. Non-linear Activation Functions


#### Linear Activation Function<br>

![](https://miro.medium.com/max/361/1*eIPQSMvZlRVxEK1Q0wwR4Q.png)

As you can see the function is a line or linear. 
Therefore, the output of the functions will not be confined between any range.<br>
Equation: f(x) = x<br>
Range: (-infinity to infinity)<br>

#### Non-linear Activation Function
![](https://miro.medium.com/max/422/1*fS2GEcT6acdgdI5nHcpN4A.png)
<br>
The Nonlinear Activation Functions are the most used activation functions. Nonlinearity helps to makes the graph look something like this<br>

It makes it easy for the model to generalize or adapt to a variety of data and to differentiate between the outputs.

The main terminologies needed to understand for nonlinear functions are:<br>
**Derivative or Differential:** Change in y-axis w.r.t. change in the x-axis. It is also known as a slope.<br>
**Monotonic function:** A function that is either entirely non-increasing or non-decreasing.

Types of Non-linear Activation Function
1. Sigmoid Activation Function
2. Tanh Activation Function
3. ReLU (Rectified Linear Unit) Activation Function


#### <u>1.Sigmoid Activation Function</u>
The Sigmoid Function curve looks like an S-shape.

![](https://miro.medium.com/max/458/1*z4Tkg2HbdueeJW9dnbX_GQ.png)

1. We use the sigmoid function is that it exists between (0 to 1). 
2.  It is especially used for models where we have to predict the probability as an output.
3. The probability of anything exists only between the range of 0 and 1.
4. The function is differentiable,we can find the slope of the sigmoid curve at any two points.
5. The function is monotonic but the function’s derivative is not.
6. The softmax function is a more generalized logistic activation function that is used for multiclass classification.

Reference:https://en.wikipedia.org/wiki/Sigmoid_function

#### <u>2. Tanh Activation Function</u>

tanh is also like logistic sigmoid but better. The range of the tanh function is from (-1 to 1). tanh is also sigmoidal (s-shaped). <br>

![](https://miro.medium.com/max/298/1*kAWj5nIwy4RZeamhXrm7ow.png)

1. The advantage is that the negative inputs will be mapped strongly negative and the zero inputs will be mapped near zero in the tanh graph.
2. The function is differentiable.
3. The function is monotonic while its derivative is not monotonic.
4. The tanh function is mainly used classification between two classes.
5. Both tanh and logistic sigmoid activation functions are used in feed-forward nets.

Reference:https://en.wikipedia.org/wiki/Activation_function




#### <u>3.ReLU (Rectified Linear Unit) Activation Function</u>

The ReLU is the most used activation function in the world right now. Since it is used in almost all the convolutional neural networks or deep learning.

![](https://miro.medium.com/max/276/1*s-d5PJ75U9GFDzfZYmvosA.png)

1. The ReLU is half rectified (from bottom).
2. f(z) is zero when z is less than zero and f(z) is equal to z when z is above or equal to zero.
3. Range: [ 0 to infinity]
4. The function and it is derivative both are monotonic.
5. The issue is that all the negative values become zero immediately which decreases the ability of the model to fit or train from the data properly. 

Reference:https://en.wikipedia.org/wiki/Rectifier_(neural_networks)

#### <u>Pratical Implementation of ANN</u>

After looking at the theory part of ANN, let's move ahead with the implementation. 
 

In this ANN Implementation,customer churn is simply the rate at which customers leave doing business with an entity. Simply put, churn prediction involves determining the possibility of customers stopping doing business with an entity. In other words, if a consumer has purchased a subscription to a particular service, we must determine the likelihood that the customer would leave or cancel the membership.
It is a critical prediction for many businesses because acquiring new clients often costs more than retaining existing ones. Customer churn measures how and why are customers leaving the business.

You download the dataset from here<br>
https://drive.google.com/file/d/1yPCwRKWMRrIPBmDZtV1rcZJXdaOb0dAU/view?usp=sharing

We will start importing <br>
**Pandas** for data analysis,<br>
**Numpy** for calculating N-dimensional array,<br>
and 
**Tensorflow** is used for multiple tasks but has a particular focus on the training and inference of deep neural networks and  Keras acts as an interface for the TensorFlow library..

In [2]:
#import libraries
import pandas as pd
import numpy as np
import tensorflow as tf

Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.
Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.


2024-06-15 21:52:43.402800: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: SSE4.1 SSE4.2, in other operations, rebuild TensorFlow with the appropriate compiler flags.


we will define our dataset and then we will see our churn dataset for overview.

In [14]:
#loading dataset using pandas
churn_data = pd.read_csv('Churn_Modelling.csv')

In [15]:
#partial view of dataset from top
churn_data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [16]:
#dimension of dataset using shape
churn_data.shape

(10000, 14)

In this dataset there are 10000 rows and 14 columns are present. There are some categorical and some numerical columns present.

Now time to preprocess the data, <br>
firstly we will observe the dataset,  this means we have to see the data types of the columns.
we will check the dataset information using the info().

In [17]:
#basic dataset information
churn_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           10000 non-null  int64  
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(2), int64(9), object(3)
memory usage: 1.1+ MB


You can see that the datatypes of each column, number of rows present with non-null values, there are many int float, and remaining are string datatype columns.



Now we will summarize the statistical part by usning describe method.

In [18]:
#basic statistics symmary
churn_data.describe()

RowNumber    CustomerId   CreditScore           Age        Tenure  \
count  10000.00000  1.000000e+04  10000.000000  10000.000000  10000.000000   
mean    5000.50000  1.569094e+07    650.528800     38.921800      5.012800   
std     2886.89568  7.193619e+04     96.653299     10.487806      2.892174   
min        1.00000  1.556570e+07    350.000000     18.000000      0.000000   
25%     2500.75000  1.562853e+07    584.000000     32.000000      3.000000   
50%     5000.50000  1.569074e+07    652.000000     37.000000      5.000000   
75%     7500.25000  1.575323e+07    718.000000     44.000000      7.000000   
max    10000.00000  1.581569e+07    850.000000     92.000000     10.000000   

             Balance  NumOfProducts    HasCrCard  IsActiveMember  \
count   10000.000000   10000.000000  10000.00000    10000.000000   
mean    76485.889288       1.530200      0.70550        0.515100   
std     62397.405202       0.581654      0.45584        0.499797   
min         0.000000       1.000000      0.00000        0.000000   
25%         0.000000       1.000000      0.00000        0.000000   
50%     97198.540000       1.000000      1.00000        1.000000   
75%    127644.240000       2.000000      1.00000        1.000000   
max    250898.090000       4.000000      1.00000        1.000000   

       EstimatedSalary        Exited  
count     10000.000000  10000.000000  
mean     100090.239881      0.203700  
std       57510.492818      0.402769  
min          11.580000      0.000000  
25%       51002.110000      0.000000  
50%      100193.915000      0.000000  
75%      149388.247500      0.000000  
max      199992.480000      1.000000

Now we have to check for null values, for this, we use the pandas IsNull() method which will give True if the null value is present and False when there are no null values.

In [19]:
#checking for missing values
print(str('Any missing data or NaN in the dataset:\n'),churn_data.isnull().sum())

Any missing data or NaN in the dataset:
 RowNumber          0
CustomerId         0
Surname            0
CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64


As we can see there is no Nan values in dataset but there some categorical features.

As there is no importance in cust id, row no and sur name for modelling we are not included here in independent feature

In [20]:
#assigning independant features without including the cust id,row no and surname.
drop_deatures = ['CustomerId', 'Surname', 'Exited', 'RowNumber']
X = churn_data.drop(drop_deatures, axis=1)
#assigning dependant feature i.e churn
Y = churn_data['Exited']

In [21]:
#independent features
X

CreditScore Geography  Gender  Age  Tenure    Balance  NumOfProducts  \
0             619    France  Female   42       2       0.00              1   
1             608     Spain  Female   41       1   83807.86              1   
2             502    France  Female   42       8  159660.80              3   
3             699    France  Female   39       1       0.00              2   
4             850     Spain  Female   43       2  125510.82              1   
...           ...       ...     ...  ...     ...        ...            ...   
9995          771    France    Male   39       5       0.00              2   
9996          516    France    Male   35      10   57369.61              1   
9997          709    France  Female   36       7       0.00              1   
9998          772   Germany    Male   42       3   75075.31              2   
9999          792    France  Female   28       4  130142.79              1   

      HasCrCard  IsActiveMember  EstimatedSalary  
0             1               1        101348.88  
1             0               1        112542.58  
2             1               0        113931.57  
3             0               0         93826.63  
4             1               1         79084.10  
...         ...             ...              ...  
9995          1               0         96270.64  
9996          1               1        101699.77  
9997          0               1         42085.58  
9998          1               0         92888.52  
9999          1               0         38190.78  

[10000 rows x 10 columns]

In [22]:
#dependent features
Y

0       1
1       0
2       1
3       0
4       0
       ..
9995    0
9996    0
9997    1
9998    1
9999    0
Name: Exited, Length: 10000, dtype: int64

As we have two columns with categorical values we go for encoding, we need to convert them into numerical values i.e Categorical encoding <br>
Gender will have some correlation with other features so it is necessary to keep that feature in the data so we go for label encoding. For that we import the LabelEncoder library from sklearn.preprocessing. Label Encoding refers to converting the labels into a numeric form so as to convert them into the machine-readable form.<br> Refer documentation for label encoding: https://www.geeksforgeeks.org/ml-label-encoding-of-datasets-in-python/. <br>
Next we create labelEncoder object for using features of converting into machine readable form. After creating object of labelEncoder We will fit and transform the gender column into it for machine readable form. 

In [28]:
#importing LabelEnoder from sklearn
from sklearn.preprocessing import LabelEncoder
#gender column in index 2
le = LabelEncoder()
X['Gender'] = le.fit_transform(X['Gender'])
#The above line is given for fit and transform of gender column

The feature country has more than two unique values, so OneHotEncoding has to be used for encoding the column. Import the libraries ColumnTransformer and OneHotEncoder. ColumnTransformer allows different columns or column subsets of the input to be transformed separately and the features generated by each transformer will be concatenated to form a single feature space. <br>
Refer documentation: https://scikit-learn.org/stable/modules/generated/sklearn.compose.ColumnTransformer.html <br>
In OneHotEncoding, each category value is converted into a new column and assigned a 1 or 0 (notation for true/false) value to the column.<br>
Refer documentation: http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html. <br>
The ct object of the ColumnTransformer class is created using the parameters transformers and remainder. transformers parameter contains the encoder name and the column index to be encoded. As country name is the second column, index is set to 1. remainder parameter has value that decides whether to include only the specified column or other columns as well in the output. 'passthrough' value is passed to remainder which ensures that only the specified column(country name) is transformed and combined in the output.

In [30]:
# Importing ColumnTransformer,OneHotEncoder 
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
#country name is present in 1st index value
cl = ColumnTransformer(transformers=[
    ('encoder',OneHotEncoder(),[1])
] ,remainder='passthrough')

X_transformed = cl.fit_transform(X)

In [46]:
# Get the column names for the OneHotEncoded columns
one_hot_encoded_columns = ct.transformers_[0][1].get_feature_names_out(['Country'])
# Combine one-hot encoded columns with the original columns (excluding the original 'Country' column)
transformed_df = pd.DataFrame(
    X_transformed,
    columns=list(one_hot_encoded_columns) + [X.columns[0]] + X.columns[2:].tolist()
)

In [47]:
transformed_df

Country_France  Country_Germany  Country_Spain  CreditScore  Gender  \
0                1.0              0.0            0.0        619.0     0.0   
1                0.0              0.0            1.0        608.0     0.0   
2                1.0              0.0            0.0        502.0     0.0   
3                1.0              0.0            0.0        699.0     0.0   
4                0.0              0.0            1.0        850.0     0.0   
...              ...              ...            ...          ...     ...   
9995             1.0              0.0            0.0        771.0     1.0   
9996             1.0              0.0            0.0        516.0     1.0   
9997             1.0              0.0            0.0        709.0     0.0   
9998             0.0              1.0            0.0        772.0     1.0   
9999             1.0              0.0            0.0        792.0     0.0   

       Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     42.0     2.0       0.00            1.0        1.0             1.0   
1     41.0     1.0   83807.86            1.0        0.0             1.0   
2     42.0     8.0  159660.80            3.0        1.0             0.0   
3     39.0     1.0       0.00            2.0        0.0             0.0   
4     43.0     2.0  125510.82            1.0        1.0             1.0   
...    ...     ...        ...            ...        ...             ...   
9995  39.0     5.0       0.00            2.0        1.0             0.0   
9996  35.0    10.0   57369.61            1.0        1.0             1.0   
9997  36.0     7.0       0.00            1.0        0.0             1.0   
9998  42.0     3.0   75075.31            2.0        1.0             0.0   
9999  28.0     4.0  130142.79            1.0        1.0             0.0   

      EstimatedSalary  
0           101348.88  
1           112542.58  
2           113931.57  
3            93826.63  
4            79084.10  
...               ...  
9995         96270.64  
9996        101699.77  
9997         42085.58  
9998         92888.52  
9999         38190.78  

[10000 rows x 12 columns]

we have to split our dataset into train and test sets, where the training set is used to train the model, and the testing set is used for testing the values of targeted columns.

In [48]:
#training and testing split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(transformed_df, Y, test_size=0.3)

We have just imported the train_test_split() method from the sklearn and we set some parameters where testing size was 30% and the remaining 70% considered as training data.

StandardScaler : It transforms the data in such a manner that it has mean as 0 and standard deviation as 1. In short, it standardizes the data. Standardization is useful for data which has negative values. It arranges the data in a standard normal distribution. It is more useful in classification than regression.<br>
Refer for Documentation:
https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html

In [51]:
#feature scaling is an important and mandatory for ann process before modelling for X train and X test .
#importing StandScaler from sklearn
from sklearn.preprocessing import StandardScaler
#creating Standscalar
scalar = StandardScaler()
#fit and transform of X train and X test
X_train_scaled = scalar.fit_transform(X_train)
X_test_scaled = scalar.transform(X_test)

In [56]:
X_train_scaled.shape

(7000, 12)

we have to define our model, which means we have to set the parameters and layers of the deep neural network which will be used for training the data.

In [52]:
#ANN - initializing
# Initializing the ANN by calling the Sequential class from keras of Tensorflow
from tensorflow.keras.models import Sequential
ann = Sequential()

we define sequential model, in the sequential model the input, hidden and output layers are connected into the sequential manner, here we define one input layer which contains all 6 columns as an input, second is hidden layers which contain 6,here we apply RelU activation function. Our last layer is the output layer, as our output is in the form of 1 and 0 so, we will use the sigmoid activation function.

In [58]:
#input layer
# 12 features
# Adding "fully connected" INPUT layer to the Sequential ANN by calling Dense class
# Number of Units = 12 and Activation Function = Rectifier
from tensorflow.keras.layers import Dense

ann.add(Dense(units=12, activation='relu', input_dim=X_train_scaled.shape[1]))

Here we have added a input layer in ann variable with 6 units and applying  RelU activation function by using dense function.

In [59]:
#hidden layer
# Adding "fully connected" SECOND layer to the Sequential ANN by calling Dense class
# Number of Units = 12 and Activation Function = Rectifier
ann.add(Dense(units=12, activation='relu'))

As we have added a hiddien with same parameter because it gives better accuary using this parameter.<br> 
You try with our units for playing with hidden layers or by adding multiple hidden layers :)

In [60]:
#output layer
#as target value is binary - AF
# Adding "fully connected" OUTPUT layer to the Sequential ANN by calling Dense class
# we use sigmoid for binary output
# Number of Units = 1 and Activation Function = Sigmoid
ann.add(Dense(units=1, activation='sigmoid'))

we compile our sequential model and fit the training data into our model.
The compilation of the model is the final step of creating an artificial neural model. The compile defines the loss function, the optimizer, and the metrics which we have to give into parameters.

Here we use compile method for compiling the model, we set some parameters into the compile method.

In [61]:
#loss - target is binary 
# Compiling the ANN
# Type of Optimizer = Adam Optimizer, Loss Function =  crossentropy for binary dependent variable, and Optimization is done w.r.t. accuracy
ann.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

now we fit our model to training data for 50 epochs and batch size=32.by using the X train and y train.

In [63]:
#training set
ann.fit(X_train_scaled, y_train, batch_size=32, epochs=50)

Epoch 1/50
219/219 [==============================] - 1s 683us/step - loss: 0.5394 - accuracy: 0.7513
Epoch 2/50
219/219 [==============================] - 0s 631us/step - loss: 0.4326 - accuracy: 0.8146
Epoch 3/50
219/219 [==============================] - 0s 634us/step - loss: 0.4126 - accuracy: 0.8269
Epoch 4/50
219/219 [==============================] - 0s 1ms/step - loss: 0.3988 - accuracy: 0.8350
Epoch 5/50
219/219 [==============================] - 0s 633us/step - loss: 0.3848 - accuracy: 0.8423
Epoch 6/50
219/219 [==============================] - 0s 632us/step - loss: 0.3728 - accuracy: 0.8490
Epoch 7/50
219/219 [==============================] - 0s 631us/step - loss: 0.3637 - accuracy: 0.8544
Epoch 8/50
219/219 [==============================] - 0s 636us/step - loss: 0.3571 - accuracy: 0.8551
Epoch 9/50
219/219 [==============================] - 0s 631us/step - loss: 0.3529 - accuracy: 0.8563
Epoch 10/50
219/219 [==============================] - 0s 649us/step - loss: 0.3499 

Now try with your epochs and batch size for better understanding -__-
---

we are ready for predication beacuse we are ready with our training model.
We will predict using X_test from trainning model.



In [69]:
#test result - prediction
y_pred = ann.predict(X_test_scaled)
y_pred = (y_pred > 0.5).astype(int)
print("Predicted values:", y_pred.flatten()[0:100])

94/94 [==============================] - 0s 1ms/step
Predicted values: [0 0 0 0 0 0 0 0 1 0 0 0 1 0 0 0 1 0 0 1 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 1
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 1 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0
 0 0 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0]


Performance Matrices this is used in the classification problems, and the customer churn is also a classification problem so we use performance metrics for checking the model behavior.

At the last, we have to predict the churn which is in the form of 0 and 1 means it was a classification problem, and the performance of the classification problem is observed with the performance metrics.

What is confusion matrix in Sklearn?<br>
A confusion matrix is a tabular summary of the number of correct and incorrect predictions made by a classifier

In [70]:
#import confusion matix and accuracy score from sklearn and find thier matrix and accuracy
#accuracy and confusion matrix
from sklearn.metrics import confusion_matrix, accuracy_score

cm = confusion_matrix(y_test, y_pred)

acc = accuracy_score(y_test, y_pred)

print('Confusion Matrix: \n', cm)
print('Accuracy Score: \n', acc)

Confusion Matrix: 
 [[2243  110]
 [ 312  335]]
Accuracy Score: 
 0.8593333333333333


You can also try with anothers set of inputs by changing the test size or adding multiple hidden layers or changing the units values for better accuracies:))

That's cool, 85% is the accuracy of our model.

We have come to an end of this project but don't stop here, try as many projects of the similar type to get a better understanding of the use cases. 
Solve the practice sheet of this project to test yourself.!!
